File: web/lib/dates.ts
/**
 * Calendar-aware date helpers used for planning duration previews.
 *
 * These mirror the authoritative logic in `backend/app/utils/dates.py` so the
 * frontend can preview an end date before the backend computes the real one.
 * The backend remains the single source of truth for stored dates.
 */

function clampDay(year: number, monthIndex: number, day: number) {
  const lastDayOfMonth = new Date(year, monthIndex + 1, 0).getDate()
  return Math.min(day, lastDayOfMonth)
}

/** Add whole calendar months to a `yyyy-mm-dd` date string (local calendar math). */
export function addMonths(date: string, months: number): string {
  if (months < 0 || !date) return date
  const [year, month, day] = date.split("-").map(Number)
  if (!year || !month || !day) return date
  const total = year * 12 + (month - 1) + months
  const targetYear = Math.floor(total / 12)
  const targetMonth = (total % 12) + 1
  const targetDay = clampDay(targetYear, targetMonth - 1, day)
  const paddedMonth = String(targetMonth).padStart(2, "0")
  const paddedDay = String(targetDay).padStart(2, "0")
  return `${targetYear}-${paddedMonth}-${paddedDay}`
}

/** End date (`yyyy-mm-dd`) of a plan that starts on `startDate` and lasts `durationMonths`. */
export function endDateFromDuration(startDate: string, durationMonths: number): string {
  if (!durationMonths || durationMonths < 1 || !startDate) return ""
  return addMonths(startDate, durationMonths)
}

/** Present a fractional month count as a friendly label (e.g. 1, 3, 6, 12). */
export function formatDurationMonths(months: number | null | undefined): string {
  if (!months) return "—"
  return months === 1 ? "1 month" : `${months} months`
}

export const DURATION_PRESETS = [
  { value: "1", label: "1 month" },
  { value: "3", label: "3 months" },
  { value: "6", label: "6 months" },
  { value: "12", label: "12 months" },
] as const

/** Today's calendar date as a `yyyy-mm-dd` string (local time). */
export function startOfTodayDateString(): string {
  const now = new Date()
  const year = now.getFullYear()
  const month = String(now.getMonth() + 1).padStart(2, "0")
  const day = String(now.getDate()).padStart(2, "0")
  return `${year}-${month}-${day}`
}

/** True when a `yyyy-mm-dd` date is strictly before today. */
export function isPastDate(dateString: string): boolean {
  if (!dateString) return true
  return dateString < startOfTodayDateString()
}

/** True when an ISO start-date has already passed (the plan has started). */
export function planHasStarted(startDateIso: string | undefined | null): boolean {
  if (!startDateIso) return false
  return startDateIso.slice(0, 10) < startOfTodayDateString()
}

File: backend/app/utils/dates.py
"""Calendar-aware date arithmetic used by plan creation and scheduling.

These helpers are the single source of truth for duration -> end-date and
period stepping. They intentionally avoid naive day-counting (e.g. 12 months
!= 365 days) and handle month-end dates, leap years and year boundaries.

Conventions
-----------
- A plan's *duration* is expressed in whole calendar months.
- ``end_date = add_months(start_date, duration)`` — e.g. a 3-month plan from
  Jan 15 ends Apr 15; a 12-month plan from Oct 1 ends Oct 1 the next year.
- The number of scheduled periods is the count of due dates inside the
  ``[start, end)`` window, so a 3-month monthly plan makes 3 payments.
"""

from datetime import datetime, timezone
import calendar


def add_months(base: datetime, months: int) -> datetime:
    """Add whole calendar months to ``base``, clamping to the month's last day.

    Example: Jan 31 + 1mo -> Feb 28 (or Feb 29 on a leap year).
    """
    if months < 0:
        raise ValueError("months must be >= 0")
    year = base.year + (base.month - 1 + months) // 12
    month = (base.month - 1 + months) % 12 + 1
    day = min(base.day, calendar.monthrange(year, month)[1])
    return base.replace(year=year, month=month, day=day)


def end_date_from_duration(start: datetime, duration_months: int) -> datetime:
    """End date of a plan that starts on ``start`` and lasts ``duration_months``."""
    return add_months(start, duration_months)


def period_count(start: datetime, end: datetime, frequency: str) -> int:
    """Number of due dates that fall inside the ``[start, end)`` window."""
    if end < start:
        return 0

    if frequency == "weekly":
        return _step_count(start, end, 7)
    if frequency == "biweekly":
        return _step_count(start, end, 14)

    # monthly / custom: one due date per calendar month
    months = (end.year - start.year) * 12 + (end.month - start.month)
    return max(months, 1)


def period_dates(start: datetime, frequency: str, periods: int) -> list[datetime]:
    """Return ``periods`` due dates beginning at ``start``."""
    if periods < 1:
        return []

    dates: list[datetime] = []
    current = start
    for _ in range(periods):
        dates.append(current)
        if frequency == "weekly":
            current = _add_days(current, 7)
        elif frequency == "biweekly":
            current = _add_days(current, 14)
        else:
            current = add_months(current, 1)
    return dates


def _step_count(start: datetime, end: datetime, days: int) -> int:
    """Number of 7/14-day due dates strictly inside ``[start, end)``."""
    diff = end - start
    total_days = diff.days + (1 if diff.seconds else 0)
    if total_days <= 0:
        return 1  # degenerate window — the start period still counts
    return total_days // days + (1 if total_days % days else 0)


def _add_days(dt: datetime, days: int) -> datetime:
    from datetime import timedelta

    return dt + timedelta(days=days)


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


def start_date_in_past(start: datetime) -> bool:
    """True when ``start``'s calendar date is before today (UTC).

    Used to reject plans created with a start date in the past and to prevent
    users from self-joining plans that have already begun.
    """
    return start.date() < datetime.now(timezone.utc).date()